File: exercise/leetcode/0048/0048.py
from typing import List

class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        self.flip1(matrix)
        self.flip2(matrix)

    
    def flip1(self, matrix: List[List[int]]):
        """斜对角线翻转"""
        n = len(matrix)
        for i in range(n):
            for j in range(n-i):
                matrix[i][j], matrix[n-1-j][n-1-i] = matrix[n-1-j][n-1-i], matrix[i][j]
    
    def flip2(self, matrix):
        n = len(matrix)
        start, end = 0, n-1
        while start < end:
            for j in range(n):
                matrix[start][j], matrix[end][j] = matrix[end][j], matrix[start][j]
            start += 1
            end -= 1

File: exercise/leetcode/0048/0048.cc
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        flip1(matrix);
        flip2(matrix);
    }

    void flip1(vector<vector<int>>& matrix) {
        int n = matrix.size();
        for (int i = 0; i <= n-1; ++i) {
            for (int j = 0; j <= n-1-i; ++j) {
                int tmp = matrix[i][j];
                    swap(matrix[i][j], matrix[n-1-j][n-1-i]);
            }
        }
    }

    void flip2(vector<vector<int>>& matrix) {
        int n = matrix.size();
        for (int i = 0; i < n / 2; ++i) {
            for (int j = 0; j < n; ++j) {
                swap(matrix[i][j], matrix[n-1-i][j]);
            }
        }
    }
};